packages-exp/auth-exp/src/platform_cordova/popup_redirect/utils.ts

```diff
@@ -0,0 +1,196 @@
+    cordova.plugins.browsertab.isAvailable(browserTabIsAvailable => {
+      if (browserTabIsAvailable) {
+        cordova.plugins.browsertab.openUrl(handlerUrl);
+      } else {
+        // TODO: Return the inappbrowser ref that's returned from the open call
+        cordova.InAppBrowser.open(
+          handlerUrl,
+          _isIOS7Or8() ? '_blank' : '_system',
+          'location=yes'
+        );
+      }
+      resolve();
+    });
+  });
+}
+
+/**
+ * Checks the configuration of the Cordova environment. This has no side effect
+ * if the configuration is correct; otherwise it throws an error with the
+ * missing plugin.
+ */
+export function _checkCordovaConfiguration(auth: Auth): void {
+  // Check all dependencies installed.
+  // https://github.com/nordnet/cordova-universal-links-plugin
+  // Note that cordova-universal-links-plugin has been abandoned.
+  // A fork with latest fixes is available at:
+  // https://www.npmjs.com/package/cordova-universal-links-plugin-fix
+  _assert(
+    typeof window?.universalLinks?.subscribe === 'function',
+    auth,
+    AuthErrorCode.INVALID_CORDOVA_CONFIGURATION,
+    {
+      missingPlugin: 'cordova-universal-links-plugin-fix'
+    }
+  );
+
+  // https://www.npmjs.com/package/cordova-plugin-buildinfo
+  _assert(
+    typeof window?.BuildInfo?.packageName !== 'undefined',
+    auth,
+    AuthErrorCode.INVALID_CORDOVA_CONFIGURATION,
+    {
+      missingPlugin: 'cordova-plugin-buildInfo'
+    }
+  );
+
+  // https://github.com/google/cordova-plugin-browsertab
+  _assert(
+    typeof window?.cordova?.plugins?.browsertab?.openUrl === 'function',
+    auth,
+    AuthErrorCode.INVALID_CORDOVA_CONFIGURATION,
+    {
+      missingPlugin: 'cordova-plugin-browsertab'
+    }
+  );
+  _assert(
+    typeof window?.cordova?.plugins?.browsertab?.isAvailable === 'function',
+    auth,
+    AuthErrorCode.INVALID_CORDOVA_CONFIGURATION,
+    {
+      missingPlugin: 'cordova-plugin-browsertab'
+    }
+  );
+
+  // https://cordova.apache.org/docs/en/latest/reference/cordova-plugin-inappbrowser/
+  _assert(
+    typeof window?.cordova?.InAppBrowser?.open === 'function',
+    auth,
+    AuthErrorCode.INVALID_CORDOVA_CONFIGURATION,
+    {
+      missingPlugin: 'cordova-plugin-inappbrowser'
+    }
+  );
+}
+
+function generateSessionId(): string {
+  const chars = [];
+  const allowedChars =
+    '1234567890abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ';
+  for (let i = 0; i < SESSION_ID_LENGTH; i++) {
+    const idx = Math.floor(Math.random() * allowedChars.length);
+    chars.push(allowedChars.charAt(idx));
+  }
+  return chars.join('');
+}
+
+/**
+ * Computes the SHA-256 of a session ID. The SubtleCrypto interface is only
+ * available in "secure" contexts, which covers Cordova (which is served on a file
+ * protocol).
+ */
+async function computeSha256(sessionId: string): Promise<string> {
+  const encoder = new TextEncoder();
+  const buf = await crypto.subtle.digest('SHA-256', encoder.encode(sessionId));
```

Comment:
In IE11, this returns `CryptoOperation` (not Promise) and `await`-ing this won't give the right result.

See: https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/digest

Reply:
FYI: `CryptoOperation` is from an earlier draft here: https://www.w3.org/TR/2013/WD-WebCryptoAPI-20130108/#cryptooperation-interface

tl;dr: we need to attach `oncomplete` / `onerror` handlers instead of `await`.

Reply:
Ya I didn't realize Cordova supported IE11. I'm going to add this as a TODO but since I can't actually test this at the moment anyway I don't want to block on IE11 support here